proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+- Title: Input/Output Shaper
+- Decision driver: Vladimir
+- Start Date: 2023-12-29
+- Proposal PR: https://github.com/deepset-ai/haystack/pull/3784/
+
+# Summary
+
+Input/Output Shaper (aka Shaper) is a new pipeline component that can invoke arbitrary, registered functions, on the
+invocation context (query, documents etc.) of a pipeline and pass the new/modified variables further down the pipeline.
+
+# Basic example
+
+In the following example, we use InputOutputShaper to add a new variable `questions` to the invocation context.
```

Comment:
What about using the shorter name: `IOShaper` ?

Reply:
Excellent!

Reply:
why use IO at all? I think this will cause confusions because people usually connect IO with disk or network read/write operations.

Reply:
Ok interesting point. So simply Shaper then?

Reply:
Yes I'd go for Shaper then.